### PR TITLE
implement COPY FROM ... RETURN SUMMARY

### PR DESCRIPTION
**Pull request by seut**
Using the ``RETURN SUMMARY``, a result set with detailed error reporting will be returned instead of just the affected rows. This will give a much more detailed insight on why some resources or rows were not being imported.

As benchmarks are showing (they have some variance), the detailed summary does not result in any significant overhead.

```
+--------------------------+--------------------+------------+-------------------------------------------------------------------------------------------------------------------------------------------------------+
| branch                   |               mean | iterations | statement                                                                                                                                             |
+--------------------------+--------------------+------------+-------------------------------------------------------------------------------------------------------------------------------------------------------+
| copy-from-return-summary | 26597.7365         |         10 | copy copytest from 'file:///Users/smu/develop/projects/crate/tmp/copyFromBenchmark/data*_noerror.json' with (shared=true)                             |
| master                   | 25943.2075         |         10 | copy copytest from 'file:///Users/smu/develop/projects/crate/tmp/copyFromBenchmark/data*_noerror.json' with (shared=true)                             |
| copy-from-return-summary | 28592.597700000002 |         10 | copy copytest_rs from 'file:///Users/smu/develop/projects/crate/tmp/copyFromBenchmark/data*_noerror.json' with (shared=true) return summary           |
| copy-from-return-summary | 30049.499099999997 |         10 | copy copytest_with_errors from 'file:///Users/smu/develop/projects/crate/tmp/copyFromBenchmark/data*_error.json' with (shared=true)                   |
| master                   | 30164.4604         |         10 | copy copytest_with_errors from 'file:///Users/smu/develop/projects/crate/tmp/copyFromBenchmark/data*_error.json' with (shared=true)                   |
| copy-from-return-summary | 28781.890499999998 |         10 | copy copytest_with_errors_rs from 'file:///Users/smu/develop/projects/crate/tmp/copyFromBenchmark/data*_error.json' with (shared=true) return summary |
+--------------------------+--------------------+------------+-------------------------------------------------------------------------------------------------------------------------------------------------------+
```

The dedicated commits will be squashed once accepted.

### CODE REVIEW

File: sql/src/main/java/io/crate/execution/engine/indexing/UpsertResults.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+                }
+            });
+        return s::iterator;
+    }
+
+    static class Result {
+
+        private static final String ERROR_COUNT_KEY = "count";
+
+        long successRowCount = 0;
```
Comment (chaudum) [inline sql/src/main/java/io/crate/execution/engine/indexing/UpsertResults.java]: all fields could be `private` as they are only accessed within class and outer class

File: sql/src/main/java/io/crate/execution/engine/indexing/UpsertResults.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        long errorRowCount = 0;
+        Map<String, Map<String, Object>> errors = new HashMap<>();
```
Comment (chaudum) [inline sql/src/main/java/io/crate/execution/engine/indexing/UpsertResults.java]: could be `final`

File: sql/src/main/java/io/crate/planner/statement/CopyStatementPlanner.java [hunk 3]
```diff
@@ -283,6 +337,7 @@ private static void addWithRefDependencies(HashSet<Symbol> toCollectUnique, Refe
     }
 
     /**
+     /**
```
Comment (chaudum) [inline sql/src/main/java/io/crate/planner/statement/CopyStatementPlanner.java]: unintentional?

File: sql/src/test/groovy/io/crate/execution/engine/collect/files/URLFileInputTest.groovy [hunk 4]
```diff
@@ -22,19 +22,21 @@
...
-class URLFileInputTest extends RandomizedTest {
-
+class URLFileInputTest extends CrateUnitTest {
 
     @Test
     void testGetStream() {
-        Path tempDir = newTempDir();
-        File file = new File("dont_exists", tempDir.toFile())
+        Path tempDir = createTempDir();
+        File file = new File(tempDir.toFile(), "dont_exists")
```
Comment (chaudum) [inline sql/src/test/groovy/io/crate/execution/engine/collect/files/URLFileInputTest.groovy]: grammer: `doesnt_exist`

File: sql/src/main/java/io/crate/exceptions/SQLExceptions.java [hunk 5]
```diff
@@ -223,4 +223,12 @@ public static boolean isDocumentAlreadyExistsException(Throwable e) {
         return e instanceof VersionConflictEngineException
                    && e.getMessage().contains("document already exists");
     }
+
+    /**
+     * Converts a possible ES exception to a Crate one and returns the message.
+     * The message will not contain any information about possible nested exceptions.
+     */
+    public static String userFriendlyCrateExceptionTopOnly(Throwable e) {
```
[the remark this answers is not in the document]
Comment (seut) [reply sql/src/main/java/io/crate/exceptions/SQLExceptions.java]: afaik the comment is describing what "TopOnly" means well. it's is not obvious that a CrateException is user friendly, so imho this new name describes the intention much better. Any suggestions for replacing "TopOnly"?

File: shared/src/main/java/io/crate/exceptions/Exceptions.java [hunk 6]
```diff
@@ -47,7 +47,7 @@ public static void rethrowUnchecked(final Throwable ex) {
      * Does not contain a proper stacktrace.
      * @return a String of format ExceptionName[msg];...
      */
-    public static String userFriendlyMessage(Throwable t) {
+    public static String userFriendlyMessageInclNested(Throwable t) {
```
[the remark this answers is not in the document]
Comment (seut) [reply shared/src/main/java/io/crate/exceptions/Exceptions.java]: also here, all is about user friendly messages...

**Pull request merged**